libraries/AP_Generator/AP_Generator_RichenPower.cpp

```diff
@@ -146,9 +144,54 @@ bool AP_Generator_RichenPower::get_reading()
 
     body_length = 0;
 
+    // update the time we started idling at:
+    if (last_reading.mode == Mode::IDLE) {
+        if (idle_state_start_ms == 0) {
+            idle_state_start_ms = last_reading_ms;
+        }
+    } else {
+        idle_state_start_ms = 0;
+    }
+
     return true;
 }
 
+void AP_Generator_RichenPower::update_heat()
```

Comment:
It would be good to have some comments above all functions.  I don't think it will be clear to most readers that this is a generator heat estimator.